File: glean-core/csharp/GleanTests/Metrics/StringListMetricTypeTest.cs
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

using Mozilla.Glean.Testing;
using System;
using System.IO;
using Xunit;
using static Mozilla.Glean.Glean;

namespace Mozilla.Glean.Tests.Metrics
{
    public class StringListMetricTypeTest
    {
        public StringListMetricTypeTest()
        {
            // Get a random test directory just for this single test.
            string tempDataDir = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName());

            // In xUnit, the constructor will be called before each test. This
            // feels like a natural place to initialize / reset Glean.
            GleanInstance.Reset(
                applicationId: "org.mozilla.csharp.tests",
                applicationVersion: "1.0-test",
                uploadEnabled: true,
                configuration: new Configuration(),
                dataDir: tempDataDir
                );
        }

        [Fact]
        public void APISavesToStorageByAddingThenSetting()
        {
            // Define a 'stringListMetric' string list metric, which will be stored in "store1".
            var stringListMetric = new Private.StringListMetricType(
                category: "telemetry",
                disabled: false,
                lifetime: Private.Lifetime.Application,
                name: "string_list_metric",
                sendInPings: new string[] { "store1" }
            );

            // Record two lists using add and set.
            stringListMetric.Add("value1");
            stringListMetric.Add("value2");
            stringListMetric.Add("value3");

            // Check that data was properly recorded.
            var snapshot = stringListMetric.TestGetValue();
            Assert.Equal(3, snapshot.Length);
            Assert.True(stringListMetric.TestHasValue());
            Assert.Equal("value1", snapshot[0]);
            Assert.Equal("value2", snapshot[1]);
            Assert.Equal("value3", snapshot[2]);

            // Use Set() to see that the first list is replaced by the new list.
            stringListMetric.Set(new string[]{ "other1", "other2", "other3"});
            // Check that data was properly recorded.
            var snapshot2 = stringListMetric.TestGetValue();
            Assert.Equal(3, snapshot2.Length);
            Assert.True(stringListMetric.TestHasValue());
            Assert.Equal("other1", snapshot2[0]);
            Assert.Equal("other2", snapshot2[1]);
            Assert.Equal("other3", snapshot2[2]);
        }

        [Fact]
        public void APISavesToStorageBySettingThenAdding()
        {
            // Define a 'stringListMetric' string list metric, which will be stored in "store1".
            var stringListMetric = new Private.StringListMetricType(
                category: "telemetry",
                disabled: false,
                lifetime: Private.Lifetime.Application,
                name: "string_list_metric",
                sendInPings: new string[] { "store1" }
            );

            // Record two lists using set and add.
            stringListMetric.Set(new string[] { "value1", "value2", "value3" });

            // Check that data was properly recorded.
            var snapshot = stringListMetric.TestGetValue();
            Assert.Equal(3, snapshot.Length);
            Assert.True(stringListMetric.TestHasValue());
            Assert.Equal("value1", snapshot[0]);
            Assert.Equal("value2", snapshot[1]);
            Assert.Equal("value3", snapshot[2]);

            // Use Add() to see that the list is appended to.
            stringListMetric.Add("added1");
            // Check that data was properly recorded.
            var snapshot2 = stringListMetric.TestGetValue();
            Assert.Equal(4, snapshot2.Length);
            Assert.True(stringListMetric.TestHasValue());
            Assert.Equal("value1", snapshot2[0]);
            Assert.Equal("value2", snapshot2[1]);
            Assert.Equal("value3", snapshot2[2]);
            Assert.Equal("added1", snapshot2[3]);
        }

        [Fact]
        public void DisabledListMustNotRecordData()
        {
            // Define a 'stringListMetric' string list metric, which will be stored in "store1".
            // It's disabled so it should not record anything.
            var stringListMetric = new Private.StringListMetricType(
                category: "telemetry",
                disabled: true,
                lifetime: Private.Lifetime.Application,
                name: "string_list_metric",
                sendInPings: new string[] { "store1" }
            );

            // Attempt to store the string list using set.
            stringListMetric.Set(new string[] { "value1", "value2", "value3" });
            // Check that nothing was recorded.
            // StringLists must not be recorded if they are disabled.
            Assert.False(stringListMetric.TestHasValue());

            // Attempt to store the string list using add.
            stringListMetric.Add("value4");
            // Check that nothing was recorded.
            // StringLists must not be recorded if they are disabled.
            Assert.False(stringListMetric.TestHasValue());
        }

        [Fact]
        public void TestGetValueThrows()
        {
            var stringListMetric = new Private.StringListMetricType(
                disabled: true,
                category: "telemetry",
                lifetime: Private.Lifetime.Application,
                name: "string_list_metric",
                sendInPings: new string[] { "store1" }
            );
            Assert.Throws<NullReferenceException>(() => stringListMetric.TestGetValue());
        }

        [Fact]
        public void APISavesToSecondaryPings()
        {
            // Define a 'stringListMetric' string list metric, which will be stored in "store1".
            var stringListMetric = new Private.StringListMetricType(
                category: "telemetry",
                disabled: false,
                lifetime: Private.Lifetime.Application,
                name: "string_list_metric",
                sendInPings: new string[] { "store1", "store2" }
            );

            // Record two lists using Add() and Set().
            stringListMetric.Add("value1");
            stringListMetric.Add("value2");
            stringListMetric.Add("value3");

            // Check that data was properly recorded in the second ping.
            Assert.True(stringListMetric.TestHasValue("store2"));
            var snapshot = stringListMetric.TestGetValue("store2");
            Assert.Equal(3, snapshot.Length);
            Assert.Equal("value1", snapshot[0]);
            Assert.Equal("value2", snapshot[1]);
            Assert.Equal("value3", snapshot[2]);

            // Use Set() to see that the first list is replaced by the new list.
            stringListMetric.Set(new string[] { "other1", "other2", "other3" });
            // Check that data was properly recorded in the second ping.
            Assert.True(stringListMetric.TestHasValue("store2"));
            var snapshot2 = stringListMetric.TestGetValue("store2");
            Assert.Equal(3, snapshot2.Length);
            Assert.Equal("other1", snapshot2[0]);
            Assert.Equal("other2", snapshot2[1]);
            Assert.Equal("other3", snapshot2[2]);
        }

        [Fact]
        public void LongStringListsAreTruncated()
        {
            // Define a 'stringListMetric' string list metric, which will be stored in "store1".
            var stringListMetric = new Private.StringListMetricType(
                disabled: false,
                category: "telemetry",
                lifetime: Private.Lifetime.Application,
                name: "string_list_metric",
                sendInPings: new string[] { "store1" }
            );

            for (int x = 0; x <= 20; x++)
            {
                stringListMetric.Add("value" + x);
            }

            var snapshot = stringListMetric.TestGetValue("store1");
            Assert.Equal(20, snapshot.Length);

            Assert.Equal(1, stringListMetric.TestGetNumRecordedErrors(ErrorType.InvalidValue));
        }
    }
}

File: glean-core/android/src/test/java/mozilla/telemetry/glean/private/StringListMetricTypeTest.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
* file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/* This file is based on the tests in the Glean android-components implentation.
 *
 * Care should be taken to not reorder elements in this file so it will be easier
 * to track changes in Glean android-components.
 */

package mozilla.telemetry.glean.private

import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import mozilla.telemetry.glean.testing.ErrorType
import mozilla.telemetry.glean.testing.GleanTestRule
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class StringListMetricTypeTest {

    @get:Rule
    val gleanRule = GleanTestRule(ApplicationProvider.getApplicationContext())

    @Test
    fun `The API saves to its storage engine by first adding then setting`() {
        // Define a 'stringListMetric' string list metric, which will be stored in "store1"
        val stringListMetric = StringListMetricType(
            CommonMetricData(
                disabled = false,
                category = "telemetry",
                lifetime = Lifetime.APPLICATION,
                name = "string_list_metric",
                sendInPings = listOf("store1")
            )
        )

        // Record two lists using add and set
        stringListMetric.add("value1")
        stringListMetric.add("value2")
        stringListMetric.add("value3")

        // Check that data was properly recorded.
        val snapshot = stringListMetric.testGetValue()!!
        assertEquals(3, snapshot.size)
        assertTrue(stringListMetric.testHasValue())
        assertEquals("value1", snapshot[0])
        assertEquals("value2", snapshot[1])
        assertEquals("value3", snapshot[2])

        // Use set() to see that the first list is replaced by the new list
        stringListMetric.set(listOf("other1", "other2", "other3"))
        // Check that data was properly recorded.
        val snapshot2 = stringListMetric.testGetValue()!!
        assertEquals(3, snapshot2.size)
        assertTrue(stringListMetric.testHasValue())
        assertEquals("other1", snapshot2[0])
        assertEquals("other2", snapshot2[1])
        assertEquals("other3", snapshot2[2])
    }

    @Test
    fun `The API saves to its storage engine by first setting then adding`() {
        // Define a 'stringListMetric' string list metric, which will be stored in "store1"
        val stringListMetric = StringListMetricType(
            CommonMetricData(
                disabled = false,
                category = "telemetry",
                lifetime = Lifetime.APPLICATION,
                name = "string_list_metric",
                sendInPings = listOf("store1")
            )
        )

        // Record two lists using set and add
        stringListMetric.set(listOf("value1", "value2", "value3"))

        // Check that data was properly recorded.
        val snapshot = stringListMetric.testGetValue()!!
        assertEquals(3, snapshot.size)
        assertTrue(stringListMetric.testHasValue())
        assertEquals("value1", snapshot[0])
        assertEquals("value2", snapshot[1])
        assertEquals("value3", snapshot[2])

        // Use add() to see that the list is appended to
        stringListMetric.add("added1")
        // Check that data was properly recorded.
        val snapshot2 = stringListMetric.testGetValue()!!
        assertEquals(4, snapshot2.size)
        assertTrue(stringListMetric.testHasValue())
        assertEquals("value1", snapshot2[0])
        assertEquals("value2", snapshot2[1])
        assertEquals("value3", snapshot2[2])
        assertEquals("added1", snapshot2[3])
    }

    @Test
    fun `disabled lists must not record data`() {
        // Define a string list metric which will be stored in "store1".
        // It's disabled so it should not record anything.
        val stringListMetric = StringListMetricType(
            CommonMetricData(
                disabled = true,
                category = "telemetry",
                lifetime = Lifetime.APPLICATION,
                name = "string_list_metric",
                sendInPings = listOf("store1")
            )
        )

        // Attempt to store the string list using set.
        stringListMetric.set(listOf("value1", "value2", "value3"))
        // Check that nothing was recorded.
        assertFalse(
            "StringLists must not be recorded if they are disabled",
            stringListMetric.testHasValue()
        )

        // Attempt to store the string list using add.
        stringListMetric.add("value4")
        // Check that nothing was recorded.
        assertFalse(
            "StringLists must not be recorded if they are disabled",
            stringListMetric.testHasValue()
        )
    }

    @Test
    fun `testGetValue() throws NullPointerException if nothing is stored`() {
        val stringListMetric = StringListMetricType(
            CommonMetricData(
                disabled = true,
                category = "telemetry",
                lifetime = Lifetime.APPLICATION,
                name = "string_list_metric",
                sendInPings = listOf("store1")
            )
        )
        assertNull(stringListMetric.testGetValue())
    }

    @Test
    fun `The API saves to secondary pings`() {
        // Define a 'stringListMetric' string list metric, which will be stored in "store1" and "store2"
        val stringListMetric = StringListMetricType(
            CommonMetricData(
                disabled = false,
                category = "telemetry",
                lifetime = Lifetime.APPLICATION,
                name = "string_list_metric",
                sendInPings = listOf("store1", "store2")
            )
        )

        // Record two lists using add and set
        stringListMetric.add("value1")
        stringListMetric.add("value2")
        stringListMetric.add("value3")

        // Check that data was properly recorded in the second ping.
        assertTrue(stringListMetric.testHasValue("store2"))
        val snapshot = stringListMetric.testGetValue("store2")!!
        assertEquals(3, snapshot.size)
        assertEquals("value1", snapshot[0])
        assertEquals("value2", snapshot[1])
        assertEquals("value3", snapshot[2])

        // Use set() to see that the first list is replaced by the new list.
        stringListMetric.set(listOf("other1", "other2", "other3"))
        // Check that data was properly recorded in the second ping.
        assertTrue(stringListMetric.testHasValue("store2"))
        val snapshot2 = stringListMetric.testGetValue("store2")!!
        assertEquals(3, snapshot2.size)
        assertEquals("other1", snapshot2[0])
        assertEquals("other2", snapshot2[1])
        assertEquals("other3", snapshot2[2])
    }

    @Test
    fun `Long string lists are truncated`() {
        // Define a 'stringListMetric' string list metric, which will be stored in "store1"
        val stringListMetric = StringListMetricType(
            CommonMetricData(
                disabled = false,
                category = "telemetry",
                lifetime = Lifetime.APPLICATION,
                name = "string_list_metric",
                sendInPings = listOf("store1")
            )
        )

        for (x in 0..20) {
            stringListMetric.add("value$x")
        }

        val snapshot = stringListMetric.testGetValue("store1")!!
        assertEquals(20, snapshot.size)

        assertEquals(1, stringListMetric.testGetNumRecordedErrors(ErrorType.INVALID_VALUE))
    }
}
